OpenRA.Mods.Common/Effects/LineImpactProjectile.cs

```diff
@@ -0,0 +1,271 @@
+
+			// If this projectile is blockable, first check if any of the hit actors will block this.
+			if (info.Blockable)
+			{
+				var wasBlocked = false;
+				foreach (var victim in hitActors)
+				{
+					if (victim.TraitsImplementing<BlocksProjectiles>().Any(Exts.IsTraitEnabled))
+					{
+						// Find where on the line this actor blocked the beam, and then limit the beam to that point.
+						var projectedPoint = WorldUtils.MinimumPointLineProjection(tailPos, headPos, victim.CenterPosition);
+						if ((projectedPoint - tailPos).Length < (headPos - tailPos).Length)
+						{
+							headPos = projectedPoint;
+							wasBlocked = true;
+						}
+					}
+				}
+
+				if (wasBlocked)
+				{
+					target = headPos;
+					length = Math.Min(headTicks, length);
+					hitActors = world.FindActorsOnLine(headPos, tailPos, info.BeamWidth, info.TargetExtraSearchRadius);
+				}
+			}
+
+			if (headTicks % info.TicksBetweenImpacts == 0)
+			{
+				var falloff = 100;
+				if (hitActors.Any())
+				{
+					foreach (var victim in hitActors)
+					{
+						var currentPos = victim.CenterPosition;
+						var distFromSource = (args.Source - currentPos).Length;
+						var distFromTarget = (target - currentPos).Length;
+
+						// Current falloff multiplier = 100 - ((Current distance - Falloff Start Range) / (Falloff End Range - Falloff Start Range) * Max Falloff);
+						// With the distance limited between 0 and falloff range.
+						var falloffRange = Math.Max((info.FalloffEndDistance - info.FalloffStartDistance).Length, 1);
+						falloff = (currentPos - args.Source).Length - info.FalloffStartDistance.Length;
+						falloff = Math.Min(falloff, falloffRange);
+						falloff = Math.Max(falloff, 0);
+						falloff = 100 - (falloff * info.FalloffMaxReduction / falloffRange);
+						falloff = Math.Max(Math.Min(falloff, 100), 0);
+
+						// Only allow impacts if far enough from source, and close enough to target.
+						if (distFromSource > info.MinRangeSource.Length && distFromTarget < info.MaxTargetRange.Length)
+							Impact(victim, falloff);
+					}
+				}
+			}
+
+			var beamDone = !headTravelling && (headTicks >= length) && !tailTravelling && (tailTicks >= length);
+			if (beamDone)
+				world.AddFrameEndTask(w => w.Remove(this));
+		}
+
+		public IEnumerable<IRenderable> Render(WorldRenderer wr)
+		{
+			if (wr.World.FogObscures(tailPos) &&
+				wr.World.FogObscures(headPos))
+				yield break;
+
+			var beamDone = !headTravelling && (headTicks >= length) && !tailTravelling && (tailTicks >= length);
+
+			if (!beamDone)
+			{
+				var rc = Color.FromArgb(255, color);
+				yield return new BeamRenderable(headPos, 0, tailPos - headPos, info.BeamRenderWidth, rc);
+			}
+		}
+
+		void Impact(Actor victim, int falloffFactor)
+		{
+			var adjustedModifiers = new List<int>();
+			for (var i = 0; i < args.DamageModifiers.Count(); i++)
+				adjustedModifiers.Add(falloffFactor * args.DamageModifiers.ElementAt(i) / 100);
```

Comment:
I'm attempting to make the warhead impacts (which occur at the position of actors intersected by the beam) do less damage the further away they are from the source actor of the beam (i.e. a sonic projectile can be made to do less damage the further it is away from the sonic tank).

Adding the falloffFactor to the end of the damage modifiers list will extend the blast radius of the impact, and not affect the damage done at the center position of the hit actor.